### ISSUE TITLE
SDI-12 and Teensy 3.2

### ISSUE THREAD

**Issue by MungoBerry** [duff2013/SDI12_T3#2]
Hi,

I'm trying to retrieve measurement from a SDI-12 device.
The instrument is directly connected to Teensy.
The example seems to hangs.
I simply add a Serial.print as first line on loop func, but it never show up.
If I remove the SDI12 constructor from code, the Serial.print is displayed.
I've tested it with Serial1 and Serial3.

Any hint ?

Thank you.

**Comment by duff2013** [comment]
What SDI-12 device are you using and can you provide a small example code of the problem?

**Comment by MungoBerry** [comment]
Hi,

I'm using [DeltaOHM HD3910.1](http://www.deltaohm.com/ver2012/download/HD3910_D_uk.pdf)
The code is adapted from example:

```
#include <SDI12.h>

SDI12 soil0(&Serial3, '0', true);

void setup() {
  Serial.begin(9600);
}

void loop() {
  char data[75];
  char debug[10];
  bool error;
  
  Serial.println("THIS STRING WILL NOT SHOWN");
    
  //------------------------------transparent--------------------------------------
  memset( data, 0 ,75 );
  soil0.transparent( "0M!", data );
  Serial.print(data);

  memset( data, 0 ,75 );
  soil0.transparent( "0D0!", data );
  Serial.print(data);
    
  //-----------------------------verification--------------------------------------
  memset( data, 0, 75 );
  memset( debug, 0, 10 );
  error = soil0.verification( debug );
  if ( !error ) Serial.print( debug );

  error = soil0.returnMeasurement( data, 0 );
  if ( !error ) Serial.print( data );
    
  //---------------------------------identification--------------------------------
  memset( data, 0, 75 );
  error = soil0.identification( data );
  if (!error) Serial.print( data );
    
  //---------------------------------measurement-----------------------------------
  memset( data, 0, 75 );
  memset( debug, 0, 10 );
  error = soil0.measurement( debug );
  if ( !error ) Serial.print( debug );

  error = soil0.returnMeasurement( data, 0 );
  if (!error) Serial.print( data );
  
  delay(1000);
}
```

I'm using Arduno IDE 1.8.5 and Teensyduino 1.42.

**Comment by duff2013** [comment]
Thanks for the code, I'm still trying to find my SDI-12 test board its some where:) I only have the Decagon 5TE sensors but I don't think that will make a difference here.

**Comment by duff2013** [comment]
Haven't forgot about this:)

**Comment by duff2013** [comment]
Can you try this now, must have the latest teensyduino installed.

**Comment by MungoBerry** [comment]
Thank you.
I was able to print something to serial monitor but I'm unable to get measurement to Serial1.
I will try to power sdi-12 sensor externally.

**Comment by duff2013** [comment]
The example you gave uses Serial3? Also that sensor needs 6-30Vdc power supply for SDI12.

**Comment by MungoBerry** [comment]
No, I use Serial1, but I've already tested with Serial3 (before your commit).
Is it normal if I cannot "speak" with sensor due low power, program hangs ?

**Comment by duff2013** [comment]
It depends on the sensors electronics, do you have the ability to probe Serial1 TX line to see the waveform? Have you used this sensor with other data-loggers before? You can try the queryAddress command to see if you get response too.
```
int address;

// Used to see what your sensor address actually is.
// Only one sensor can be connected to bus at a time. 
address = soil0.queryAddress( );
if(address != -1) {
    Serial.print( "Sensor Address is " );
    Serial.println( (char)address )
}

**Comment by MungoBerry** [comment]
Hi,

sorry for the loooong delay.
The function queryAddress hangs inside send_command at line 730:

`do { yield( ); } while ( !( UART2_S1 & UART_S1_IDLE ) );            // wait for i/o to become idle`

The sensor works with arduino and I've powered it externally.
Using an oscilloscope I see the signal from Teensy but not response from sensor.

**Comment by duff2013** [comment]
I'll have to get out my setup for this probably take two weeks.

**Comment by MungoBerry** [comment]
Thank you for your time.
I moved from Serial1 to Serial3 and now I'm able to read values from the sensor.
I don't know why Serial1 do not works!

Thank you again!

**Comment by duff2013** [comment]
Ok I'll take a look at Serial1 with the latest Teensyduino installed.
